tensorboard/webapp/widgets/line_chart_v2/lib/threejs_coordinator.ts

```diff
@@ -0,0 +1,43 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+import * as THREE from 'three';
+
+import {Coordinator} from './coordinator';
+import {Rect} from './types';
+
+export class ThreeCoordinator extends Coordinator {
+  private readonly camera = new THREE.OrthographicCamera(
+    0,
+    1000,
+    1000,
+    0,
+    0,
+    100
+  );
+
+  setDomContainerRect(rect: Rect) {
+    super.setDomContainerRect(rect);
+    this.camera.left = rect.x;
+    this.camera.right = rect.x + rect.width;
+    this.camera.top = rect.y;
+    this.camera.bottom = rect.y + rect.height;
+    this.camera.updateProjectionMatrix();
+  }
+
+  getCamera() {
```

Comment:
>  simpler to have only 1 Coordinator variant, used by both all (SVG and WebGL), as isolating THREE.js-specific code/constructs in fewer files is preferable to being spread out.

I am a bit confused by the sentence but let me try to address few things that I think you are asking.

- coordinator vs. renderer: currently, two classes have very distinct responsibilities. renderer is responsible for knowing how to manipulate 3d scene or DOM to draw shapes while coordinator knows how to convert a coordinate systems while remembering when it was last updated so whoever converts the coordinate system can know whether their converted data is stale or not (this address your other question; both coordinator and consumer both will have to remember some checksum/hash/timestamp/increment/uniqueId to minimize the work).

All these different pieces of the library is actually not used by the consumer on the Angular side. At the Angular side, you will simply interact with the `Chart` class which stitches all of these and invoke the methods (please TAL at this [impl](https://github.com/tensorflow/tensorboard/blob/swl_gpu_7/tensorboard/webapp/widgets/line_chart_v2/lib/chart.ts). As such, I am not leaving too much comments with very firm API contracts. However, I can improve some comments for future readers perhaps in README.md.

Reply:
Thanks for the link to the sneak peek!

I wonder if there's any way to avoid introducing a Three-specific subclass.  The reason it exists is to update a camera used only by ThreeRenderer, since the renderer isn't aware of when the dom container changes.

`chart.ts`'s `resize()` is the only place that ever calls `this.coordinator.setDomContainerDimension(dim)`.  It does so just before calling `renderer.onResize()` [1].  If resizing is the only operation that can change the DOM container dimension, can we move this THREE.Camera update logic into `ThreeRenderer`, and have it do the camera update?
```ts
// ThreeRenderer
onResize() {
  const rect = this.coordinator.getDomContainerRect();
  this.camera.left = ...
}
```

or alternatively, if the ThreeRenderer already has a reference to its canvas element, it can keep track of whether the canvas size has changed, itself?  Then, we could get rid of `get/setDomContainerRect` completely.

[1] https://github.com/tensorflow/tensorboard/blob/swl_gpu_7/tensorboard/webapp/widgets/line_chart_v2/lib/chart.ts#L80

Reply:
Improved the ThreeCoordinator to convey why it needs to be of its own thing. Hope the revised implementation and tests are sufficient in convincing you that we should encapsulate some details of our coordinate system in its own module instead.